mne/time_frequency/_stft.py

```diff
@@ -1,10 +1,140 @@
+        (default: window_size/2).
+    %(tfr_average)s
+    %(verbose)s
+
+    Returns
+    -------
+    power : AverageTFR
+        The averaged power.
+    """
+    from ..epochs import BaseEpochs
+    from .tfr import EpochsTFR
+
+    # verbose dec is used b/c subfunctions are verbose
+    data = _get_data(inst, False)
+    picks = _pick_data_channels(inst.info)
+    info = pick_info(inst.info, picks)
+    data = data[:, picks, :]
+
+    # compute STFT on the numpy array
+    power, freqs, times = tfr_array_stft(
+        data, sfreq=info['sfreq'], window_size=window_size,
+        step_size=step_size, return_times=True,
+        verbose=verbose)
+
+    if average:
+        # how many epochs are we averaging
+        nave = len(data)
+        power = np.mean(power, axis=0)
+        out = AverageTFR(info, power, times, freqs, nave, method='stft-power')
+    else:
+        if isinstance(inst, BaseEpochs):
+            meta = deepcopy(inst._metadata)
+            evs = deepcopy(inst.events)
+            ev_id = deepcopy(inst.event_id)
+        else:
+            # if the input is of class Evoked
+            meta = evs = ev_id = None
+        out = EpochsTFR(info, power, times, freqs, method='stft-power',
+                        events=evs, event_id=ev_id, metadata=meta)
+
+    return out
+
+
+def tfr_array_stft(data, sfreq, window_size, step_size=None,
+                   return_times=False, verbose=True):
+    """STFT Short-Term Fourier Transform using a sine window.
+
+    The transformation is designed to be a tight frame that can be
+    perfectly inverted. It only returns the positive frequencies.
+
+    Parameters
+    ----------
+    x : array, shape (n_epochs, n_channels, n_times)
+        Containing multi-channels signal.
+    window_size : int
+        Length of the STFT window in samples (must be a multiple of 4).
+    step_size : int
+        Step between successive windows in samples (must be a multiple of 2,
+        a divider of ``window_size`` and smaller than ``window_size``/2)
+        (default: window_size/2).
+    return_times : bool
+        Whether to return the time points of the STFT time-frequency
+        representation. Default is False.
+    %(n_jobs)s
+    %(verbose)s
+
+    Returns
+    -------
+    X : array, shape (n_epochs, n_channels, wsize // 2 + 1, n_step)
+        STFT coefficients for positive frequencies with
+        ``n_step = ceil(T / tstep)``.
+    freqs : array, shape (wsize // 2 + 1,)
+        STFT frequencies corresponding.
+
+    See Also
+    --------
+    istft
+    stftfreq
+    """
+    _validate_type(data, np.ndarray, 'data')
+    if data.ndim != 3:
+        raise ValueError(
+            'data must be 3D with shape (n_epochs, n_channels, n_times), '
+            f'got {data.shape}')
+
+    n_epochs, n_channels, n_signals = data.shape
+
+    power_list = []
+
+    for epochidx in range(n_epochs):
+        # compute STFT
+        power = _stft(data[epochidx, ...], window_size=window_size,
```

Comment:
I doubt the stft returns power by default. It should output complex valued data. Can you check?

also I would see if I can preallocate the power_list as a numpy array before the for loop to avoid doubling the RAM usage.